Streaming Decompressor v2

cccs-sadugas opened the pull request:
### Pull Request Overview

Changes since last PR:
- Rebased from master to include memlimit changes.
- Published PRs for decoupled features.
- Addressed Travis CI build errors.
- Addressed code review comments.

### Testing Strategy

This pull request was tested by...

- [x] Added relevant unit tests.
- [x] Added relevant end-to-end tests (such as `.lzma`, `.lzma2`, `.xz` files).

### Supporting Documentation and References

Link to previous PR: #51 

### TODO
- [x] Merge dependant PR #53.
- [x] Merge dependant PR #54.

cccs-sadugas commented:
Should I squash all the commits into a single one?

gendx commented:
> Should I squash all the commits into a single one?

It would be useful once the other two pull requests are merged.

gendx commented:
I have now merged the two related pull requests.

cccs-sadugas commented:
I have rebased and squashed. The new benchmarks are:

**master**
```
test compress_65536                  ... bench:   2,489,961 ns/iter (+/- 195,741)
test compress_empty                  ... bench:       1,464 ns/iter (+/- 267)
test compress_hello                  ... bench:       2,079 ns/iter (+/- 292)
test decompress_after_compress_65536 ... bench:   2,768,485 ns/iter (+/- 616,316)
test decompress_after_compress_empty ... bench:       3,346 ns/iter (+/- 674)
test decompress_after_compress_hello ... bench:       3,851 ns/iter (+/- 590)
test decompress_big_file             ... bench:   6,503,570 ns/iter (+/- 1,350,205)
test decompress_huge_dict            ... bench:       3,959 ns/iter (+/- 758)
```

**streaming-decompressor-v2**
```
test compress_65536                  ... bench:   2,611,232 ns/iter (+/- 467,790)
test compress_empty                  ... bench:       1,459 ns/iter (+/- 198)
test compress_hello                  ... bench:       2,128 ns/iter (+/- 401)
test decompress_after_compress_65536 ... bench:   2,569,576 ns/iter (+/- 272,391)
test decompress_after_compress_empty ... bench:       3,351 ns/iter (+/- 372)
test decompress_after_compress_hello ... bench:       4,050 ns/iter (+/- 463)
test decompress_big_file             ... bench:   6,325,453 ns/iter (+/- 727,326)
test decompress_huge_dict            ... bench:       4,162 ns/iter (+/- 760)
test decompress_stream_big_file      ... bench:   6,160,083 ns/iter (+/- 726,963)
```

cccs-sadugas commented:
> Can you add unit tests that the decompressed output is the same via streaming vs. with the direct interface?

Is there anything you were looking for in particular? All tests in `tests/stream.rs` assert the expected originally decompressed data is equal to the output of the stream decompressor.

gendx commented on tests/stream.rs:
```diff
@@ -0,0 +1,274 @@
+#[cfg(feature = "enable_logging")]
```
These tests are good. It would be nice to somehow factor in the code with `tests/lzma.rs` to avoid code duplication, and make sure that new tests are automatically picked up on both "normal" and "stream" modes.

I'm not sure how to best do that yet (having tests generic over some trait for example), so it shouldn't block this pull request, but any idea is appreciated of course.

cccs-sadugas replied:
Good idea. I made the common test functions like `round_trip` test both the non-streaming and streaming implementations and got rid of the stream.rs test module.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+{
+    /// State that was owned while the error occured
+    state: Option<State<W>>,
+    /// IO error that occured in that state
```
Is it necessarily an IO error?

cccs-sadugas replied:
Looks like it is now the crate's error type.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    error: Error,
+}
+
+impl<W> StreamStateError<W>
+where
+    W: Write,
+{
+    /// Creates a new StreamStateError
+    fn new(state: Option<State<W>>, error: Error) -> Self {
+        Self { state, error }
+    }
+}
```
Given that this is only used internally, I don't think this `new` function makes things clearer. In general `StreamStateError::new(foo, bar)` is less readable than `StreamStateError { state: foo, error: bar }`, because the function doesn't retain the field names.

cccs-sadugas replied:
Generally I like adding `new` functions to initialize all structs. This makes it easy to change initialization in the future if the struct changes.

However, I also don't mind switching it to struct initialization.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    code: u32,
+}
+
+impl<W> RunState<W>
+where
+    W: Write,
+{
+    fn new(decoder: DecoderState<W, LZCircularBuffer<W>>, range: u32, code: u32) -> Self {
+        Self {
+            decoder,
+            range,
+            code,
+        }
+    }
+}
```
Same here.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    }
+}
+
+impl<W> std::fmt::Debug for RunState<W>
```
I'd import `use std::fmt::Debug` at the file level.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    /// Initialize the stream with the given `options`. This will consume the
+    /// `output` which is the sink implementing `std::io::Write` that will
+    /// receive decompressed bytes.
+    pub fn new_with_options(options: Options, output: W) -> Self {
```
The non-stream API takes an `&Options` as input. We should be consistent.

cccs-sadugas replied:
I'll change this for consistency. Unfortunately it results in an extra clone when calling `new()`:
```rust
pub fn new(output: W) -> Self {
    Self::new_with_options(&Options::default(), output)
}

pub fn new_with_options(options: &Options, output: W) -> Self {
    Self {
        tmp: [0; MAX_TMP_LEN],
        tmp_len: 0,
        state: Some(State::Header(output)),
        options: options.clone(),
    }
}
```

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    }
+
+    /// Get a reference to the output sink
+    pub fn get_ref(&self) -> Option<&W> {
```
The name should be clearer. From the call side it's unclear a reference to what is obtained.

cccs-sadugas replied:
I changed it to `get_output()` and `get_output_mut()` to be consistent with `LZBuffer`.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    /// Get a reference to the output sink
+    pub fn get_ref(&self) -> Option<&W> {
+        match &self.state {
+            Some(State::Init(output)) => Some(&output),
```
`self.state.map(|s| match s { ... })` would be more idiomatic.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+        options: &Options,
+    ) -> Result<State<W>, StreamStateError<W>> {
+        let params = match LZMAParams::read_header(&mut input, options) {
+            Ok(params) => params,
+            Err(e) => {
+                return Err(StreamStateError::new(Some(State::Init(output)), e));
+            }
```
This pattern would be more idiomatic as `let params = read_header(...).map_err(|e| ... )?;`

cccs-sadugas replied:
This doesn't apply now that we are no longer returning a `StreamStateError`. I did change it to an `if let Ok(...)`.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+
+        let len = match input.fill_buf() {
+            Ok(val) => val,
+            Err(_) => {
```
Here as well `map_err` would be clearer.

cccs-sadugas replied:
Same as above, using `if let Ok(...)` now.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+        }
+        .len();
+
+        if len < START_BYTES {
```
Is it an error? As far as I'm aware, `fill_buf` can return as few as 1 byte at a time, but it doesn't mean that `START_BYTES` aren't available afterwards.

From the documentation (https://doc.rust-lang.org/std/io/trait.BufRead.html#tymethod.fill_buf).

> An empty buffer returned indicates that the stream has reached EOF.

It would actually be nice to have some unit tests for this streaming API with some `BufRead` implementation that returns 1 byte at a time. Although that would be inefficient, that would also test the limits of the implementation.

cccs-sadugas replied:
There were a couple of issues with this function that I revisited. Mainly, we need to distinguish between a fatal error and "need more bytes". That's why I was doing a length check before continuing. However, this length check should have been done before calling `read_header()`. Doing so was not possible because the compiler reasonably complains that `output` is being moved multiple times.

Instead of doing a length check, I added an `Error::HeaderTooShort` (one step towards #6 maybe) so that we can distinguish between errors and handle the "need more bytes" case.

I also added `into_output()` to get back the output sink that was moved into the `DecoderState`'s `LZBuffer`.

Lastly, I added a test case `test_bad_header` to show that we properly fail on an invalid header.

cccs-sadugas replied:
Regarding the unit tests, I do have a couple of tests in `stream.rs` to test resilience like this. `test_stream_chunked` will test every single chunk size from 1 on the whole input size.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+    W: Write,
+{
+    /// Stream is initialized but header values have not yet been read.
+    Init(W),
```
I'd name these states `Header` and `Data` for clarity.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+where
+    W: Write,
+{
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
```
Do we need Debug formatting in the general case, or should it be restricted to `cfg(feature = "enable_logging")` and/or `cfg(test)`?

cccs-sadugas replied:
The build fails without debug implementations:

```
error: type does not implement `std::fmt::Debug`; consider adding `#[derive(Debug)]` or a manual implementation
  --> src/decode/stream.rs:48:1
   |
48 | / pub struct Stream<W>
49 | | where
50 | |     W: Write,
51 | | {
...  |
60 | |     options: Options,
61 | | }
   | |_^
   |
note: the lint level is defined here
  --> src/lib.rs:4:9
   |
4  | #![deny(missing_debug_implementations)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: build failed
```

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+        output: W,
+        mut input: &mut R,
+        options: &Options,
+    ) -> Result<State<W>, StreamStateError<W>> {
```
To me, this `Result<State<W>, StreamStateError<W>>` doesn't describe well what's returned by this function. To me, there are 3 options:
- Success with a state,
- Need more data (isn't it a success with some state as well?),
- A real error (`crate::error::Error`).

I don't think conflating "need more data" into the `StreamStateError` is suitable. I think the `Ok` side should represent the "need more data" status - then the caller can handle that and just forward a "real error" if it happens.

cccs-sadugas replied:
Fair point. I was able to return a `crate::error::Result<State<W>>` and use state transitions to identify the case when we need more data:
- `Ok(State::Header(_))` means we need more data.
- `Ok(State::Data(_))` means we successfully transitioned states.
- `Err(_)` is a real error.

gendx commented on src/decode/stream.rs:
```diff
@@ -0,0 +1,497 @@
+where
+    W: Write,
+{
+    /// Temporary buffer to hold data while the header is being read.
+    tmp: [u8; MAX_TMP_LEN],
+    /// How many bytes of the temp buffer are in use.
+    tmp_len: usize,
```
I think this pair of fields would deserve its own helper struct, to make sure that things are properly encapsulated, and that we can't read beyond the current `tmp_len`.

cccs-sadugas replied:
I was debating this myself. Thinking of using something like std::io::Cursor to keep track of the `tmp_len` "position" in the buffer.

cccs-sadugas commented:
Thanks for the thorough comments @gendx . I pushed changes to #58.
